velox/functions/prestosql/window/AggregateWindow.cpp

```diff
@@ -0,0 +1,199 @@
+ public:
+  explicit AggregateWindowFunction(
+      const std::string& name,
+      const std::vector<exec::WindowFunctionArg>& args,
+      const TypePtr& resultType,
+      velox::memory::MemoryPool* pool)
+      : WindowFunction(resultType, pool) {
+    argTypes_.reserve(args.size());
+    argIndices_.reserve(args.size());
+    argVectors_.reserve(args.size());
+    for (const auto& arg : args) {
+      argTypes_.push_back(arg.type);
+      // TODO : Enhance code to handle constant arguments.
+      VELOX_CHECK_NULL(arg.constantValue);
+      VELOX_CHECK(arg.index.has_value());
+      argIndices_.push_back(arg.index.value());
+      argVectors_.push_back(BaseVector::create(arg.type, 0, pool_));
+    }
+    // Create an Aggregate function object to do result computation. Window
+    // function usage only requires single group aggregation for calculating
+    // the function value for each row.
+    aggregate_ = exec::Aggregate::create(
+        name, core::AggregationNode::Step::kSingle, argTypes_, resultType);
+
+    // Aggregate initialization.
+    int32_t rowSizeOffset = bits::nbytes(1);
+    singleGroupRowSize_ = rowSizeOffset + sizeof(int32_t);
+    int32_t nullOffset = 0;
+    aggregate_->setOffsets(
+        singleGroupRowSize_,
+        exec::RowContainer::nullByte(nullOffset),
+        exec::RowContainer::nullMask(nullOffset),
+        /* needed for out of line allocations */ rowSizeOffset);
+    singleGroupRowSize_ += aggregate_->accumulatorFixedWidthSize();
+
+    // Constructing the single row in the MemoryPool.
+    singleGroupRow_ = (char*)(pool_->allocate(singleGroupRowSize_));
+
+    // Constructing a vector of a single result value used for copying from
+    // the aggregate to the final result.
+    aggregateResultVector_ = BaseVector::create(resultType, 1, pool_);
+  }
+
+  ~AggregateWindowFunction() {
+    std::vector<char*> singleGroupRowVector = {singleGroupRow_};
+    aggregate_->destroy(folly::Range(singleGroupRowVector.data(), 1));
+    pool_->free(singleGroupRow_, singleGroupRowSize_);
+  }
+
+  void resetPartition(const exec::WindowPartition* partition) override {
+    partition_ = partition;
+  }
+
+  void apply(
+      const BufferPtr& /*peerGroupStarts*/,
+      const BufferPtr& /*peerGroupEnds*/,
+      const BufferPtr& frameStarts,
+      const BufferPtr& frameEnds,
+      vector_size_t resultOffset,
+      const VectorPtr& result) override {
+    int numRows = frameStarts->size() / sizeof(vector_size_t);
+    auto frameStartsVector = frameStarts->as<vector_size_t>();
+    auto frameEndsVector = frameEnds->as<vector_size_t>();
+    fillArgVectors(frameStartsVector, frameEndsVector, numRows);
+
+    auto singleGroup = std::vector<vector_size_t>{0};
+    rows_.resize(maxFrame_ + 1 - minFrame_);
+    rows_.clearAll();
+
+    for (int i = 0; i < numRows; i++) {
+      // This is a very naive algorithm.
+      // It evaluates the entire aggregation for each row by iterating over
```

Comment:
A low handing fruit could be to optimize this for the case when frameStart is constant. In this case, we can maintain a single accumulator, for each row add that row, extract current result, add another row, extract current result, etc.